pkg/reconciler/vertex/scaling/scaling.go

```diff
@@ -223,6 +215,30 @@ func (s *Scaler) scaleOneVertex(ctx context.Context, key string, worker int) err
 		s.daemonClientsCache.Add(pl.GetDaemonServiceURL(), daemonClient)
 	}
 
+	if vertex.Status.Replicas == 0 { // Was scaled to 0
+		// Non source udfs don't need to scale up and peek to check the pending messages
+		// We will only scale them up if they have any pending messages
+		if !vertex.IsASource() {
+			totalPending, err := getPendingUpstreamMessageCount(ctx, daemonClient, pl, vertex)
```

Comment:
There's already some code to query the buffer information (to get `partitionBufferLengths`, etc), could you combine them? e.g.

```golang
func getBufferInfos() (xxxx, ...) {
  // return totalPending, partitionBufferLengths, partitionAvailableBufferLengths, error
}
```

`targetAvailableBufferLength` is not in-use, can be removed.